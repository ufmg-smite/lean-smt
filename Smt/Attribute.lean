import Lean
import Std

namespace Smt.Attribute

open Lean
open Std

/-- An extension to Lean's runtime environment to support SMT attributes.
    Maintains a set of function declarations for the `smt` tactic to utilize
    while generating the SMT query. -/
abbrev SmtExtension := SimpleScopedEnvExtension Name (HashSet Name)

/-- Adds a declaration to the set of function declarations maintained by the SMT
    environment extension. -/
def addSmtEntry (d : HashSet Name) (e : Name) : HashSet Name :=
  d.insert e

initialize smtExt : SmtExtension ← registerSimpleScopedEnvExtension {
  name     := `SmtExt
  initial  := {}
  addEntry := addSmtEntry
}

/-- Throws unexpected type error. -/
def throwUnexpectedType (t : Name) (n : Name) : AttrM Unit :=
  throwError s!"unexpected type at '{n}', `{t}` expected"

/-- Validates the tagged declaration. -/
def validate (n : Name) : AttrM Unit := do
  match (← getEnv).find? n with
  | none      => throwError s!"unknown constant '{n}'"
  | some info =>
    match info.type with
    | Expr.const c .. =>
      if c != (`Smt.Transformer) then throwUnexpectedType `Smt.Transformer n
    | _               => throwUnexpectedType `Smt.Transformer n

/-- Registers an SMT attribute with the provided name and description and links
    it against `ext`. -/
def registerSmtAttr (attrName : Name) (attrDescr : String)
  : IO Unit :=
  registerBuiltinAttribute {
    name  := attrName
    descr := attrDescr
    applicationTime := AttributeApplicationTime.afterTypeChecking
    add   := fun decl stx attrKind => do
      trace[Smt.debug.attr] s!"attrName: {attrName}, attrDescr: {attrDescr}"
      trace[Smt.debug.attr] s!"decl: {decl}, stx: {stx}, attrKind: {attrKind}"
      Attribute.Builtin.ensureNoArgs stx
      validate decl
      setEnv (smtExt.addEntry (← getEnv) decl)
      trace[Smt.debug.attr]
        s!"transformers: {(smtExt.getState (← getEnv)).toList}"
    erase := fun declName => do
      let s := smtExt.getState (← getEnv)
      let s := s.erase declName
      modifyEnv fun env => smtExt.modifyState env fun _ => s
  }

initialize registerSmtAttr `Smt
             "Utilize this function to transform Lean expressions to SMT terms."

end Smt.Attribute
