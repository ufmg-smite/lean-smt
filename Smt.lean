import Smt.General
import Smt.Nat
import Smt.Prop
import Smt.Solver
import Smt.String
import Smt.Tactic
import Smt.Term
