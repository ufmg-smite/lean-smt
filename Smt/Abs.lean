/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Translator
import Mathlib.Algebra.Order.Group.Abs

namespace Smt.Abs

open Lean Expr
open Translator Term


@[smtTranslator] def replaceAbs : Translator
  | app (app (app (app (const ``abs _) _) _) _) e => do
      let tmE ← applyTranslators! e
      return appT (symbolT "abs") tmE
  | _ => return none

end Smt.Abs
