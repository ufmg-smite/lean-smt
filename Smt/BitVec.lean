/-
Copyright (c) 2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Wojciech Nawrocki
-/

import Lean
import Smt.Translator
import Smt.Data.BitVec

namespace Smt.BitVec

open Lean Expr
open Translator Term

@[smtTranslator] def replaceType : Translator
  | e@(app (const ``BitVec _) n) => do
    let n ← Meta.whnf n
    let some n ← Meta.evalNat n |>.run
      | throwError "bitvector type{indentD e}\nhas variable width"
    return mkApp2 (symbolT "_") (symbolT "BitVec") (literalT (toString n))
  | _ => return none

def mkLit (w : Nat) (n : Nat) : Term :=
  let bits := Nat.toDigits 2 n |>.take w
  literalT <| bits.foldl (init := "#b".pushn '0' (w - bits.length)) (·.push ·)

@[smtTranslator] def replaceEq : Translator
  -- TODO: we should really just support beq/bne across all types uniformly
  | app (app (const ``BEq.beq _) (app (const ``BitVec _) _)) _ => return symbolT "="
  | app (app (app (app (const ``bne _) (app (const ``BitVec _) _)) _) e) e' =>
    return appT (symbolT "not") (mkApp2 (symbolT "=") (← applyTranslators! e) (← applyTranslators! e'))
  | _ => return none

@[smtTranslator] def replaceFun : Translator
  | app (app (const ``BitVec.append _) _) _ => return symbolT "concat"
  | app (const ``BitVec.and _) _            => return symbolT "bvand"
  | app (const ``BitVec.or _) _             => return symbolT "bvor"
  | app (const ``BitVec.xor _) _            => return symbolT "bvxor"
  | e@(app (app (app (const ``BitVec.shiftLeft _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvshl") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.shiftRight _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvlshr") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.extract _) _) i) j) => do
    let i ← reduceLit i e
    let j ← reduceLit j e
    return mkApp3 (symbolT "_") (symbolT "extract") (literalT (toString i)) (literalT (toString j))
  | e@(app (const ``BitVec.zero _) w) => do
    let w ← reduceWidth w e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w 0
  | e@(app (app (const ``BitVec.ofNat _) w) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w n
  | _ => return none
where
  reduceWidth (w : Expr) (e : Expr) : TranslationM Nat := do
    let some w ← Meta.evalNat (← Meta.whnf w) |>.run
      | throwError "bitvector width{indentD w}\nis not constant in{indentD e}"
    return w
  reduceLit (n : Expr) (e : Expr) : TranslationM Nat := do
    let some n ← Meta.evalNat (← Meta.whnf n) |>.run
      | throwError "literal{indentD n}\nis not constant in{indentD e}"
    return n

end Smt.BitVec
