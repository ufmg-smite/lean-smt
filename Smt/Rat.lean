/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Batteries.Data.Rat
import Smt.Translator
import Mathlib.Data.Rat.Lemmas
namespace Smt.Rat

open Lean Expr
open Translator Term

@[smtTranslator] def replaceConst : Translator
  | const `Rat _      => return symbolT "Real"
  | const ``Rat.add _ => return symbolT "+"
  | const ``Rat.sub _ => return symbolT "-"
  | const ``Rat.neg _ => return symbolT "-"
  | const ``Rat.mul _ => return symbolT "*"
  | const ``Rat.div _ => return symbolT "div"
  | const ``Rat.lt _  => return symbolT "<"
  | const ``Rat.blt _ => return symbolT "<"
  | app (app (const ``LE.le _) (const `Rat _)) _ => return symbolT "<="
  | app (app (const ``GT.gt _) (const `Rat _)) _ => return symbolT ">"
  | app (app (const ``GE.ge _) (const `Rat _)) _ => return symbolT ">="
  | app (app (app (const ``Int.cast _) (const `Rat _)) _) e => applyTranslators! e
  | app (app (app (const ``Rat.cast _) (const `Real _)) _) e  => applyTranslators! e
  | _ => return none

@[smtTranslator] def replaceOfScientific : Translator
  | app (app (app (const ``Rat.ofScientific _) m) _) e => do
      let tmE ← applyTranslators! e
      let tmM ← applyTranslators! m
      return Term.mkApp2 (symbolT "div") tmM (Term.mkApp2 (symbolT "^") (literalT "10") tmE)
  | _ => return none

end Smt.Rat
