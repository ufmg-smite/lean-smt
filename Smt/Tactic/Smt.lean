/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean

import Smt.Dsl.Sexp
import Smt.Query
import Smt.Solver

namespace Smt

open Lean
open Lean.Elab
open Lean.Elab.Tactic

initialize
  registerTraceClass `smt.debug
  registerTraceClass `smt.debug.attr
  registerTraceClass `smt.debug.query
  registerTraceClass `smt.debug.transformer

/-- `smt` converts the current goal into an SMT query and checks if it is
satisfiable. By default, `smt` generates the minimum valid SMT query needed to
assert the current goal. However, that is not always enough:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt
```
For the theorem above, `smt` generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(check-sat)
```
which is missing the hypotheses `hp` and `hpq` required to prove the theorem. To
pass hypotheses to the solver, use `smt [h₁, h₂, ..., hₙ]` syntax:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt [hp, hpq]
```
The tactic then generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(declare-const p Bool)
(assert p)
(assert (=> p q))
(check-sat)
```
-/
syntax (name := smt) "smt" ("[" ident,* "]")? : tactic

def queryToString (commands : List String) : String :=
  String.intercalate "\n" ("(check-sat)\n" :: commands).reverse

def parseTactic : Syntax → TacticM (List Expr)
  | `(tactic| smt)              => pure []
  | `(tactic| smt [ $[$hs],* ]) => hs.toList.mapM (fun h => elabTerm h.raw none)
  | _                           => throwUnsupportedSyntax

@[tactic smt] def evalSmt : Tactic := fun stx => withMainContext do
  -- 1. Get the current main goal.
  let goalType ← Tactic.getMainTarget
  let goalId ← Lean.mkFreshMVarId
  Lean.Meta.withLocalDeclD goalId.name (mkNot goalType) fun g => do
  -- 2. Get the hints passed to the tactic.
  let mut hs ← parseTactic stx
  hs := hs.eraseDups
  -- 3. Generate the SMT query.
  let mut solver := Solver.mk []
  solver ← Query.generateQuery g hs solver
  let query := queryToString solver.commands
  -- 4. Run the solver.
  let kind := smt.solver.kind.get (← getOptions)
  let path := smt.solver.path.get? (← getOptions) |>.getD (toString kind)
  -- 5. Print the result.
  let res ← solver.checkSat kind path
  logInfo m!"goal: {goalType}\n\nquery:\n{query}\nresult: {res}"
  let out ← match Sexp.parse res with
    | .ok out => pure out
    | .error e => throwError "cannot parse solver output: {e}"
  if out.contains sexp!{sat} then
    throwError "unable to prove goal, it is false"
    -- TODO ask for countermodel and print
  else if !out.contains sexp!{unsat} then
    throwError "unable to prove goal"

end Smt
