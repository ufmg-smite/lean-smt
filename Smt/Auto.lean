/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Auto.Tactic
import Smt.Tactic.Smt

open Lean in
def smtSolverFunc (ls : Array Auto.Lemma) (is : Array Auto.Lemma) : MetaM Expr := do
  let fi l := do
    let userName ← mkFreshUserName `inst
    let type ← Meta.mkAppM ``Inhabited #[l.type]
    let value ← Meta.mkAppOptM ``Inhabited.mk #[l.type, l.proof]
    return { userName, type, value }
  let is ← is.mapM fi
  let fl l := do
    let userName ← mkFreshUserName `h
    let type ← Auto.Lam2D.approxSimpNF l.type
    let value := l.proof
    return { userName, type, value }
  let ls ← ls.mapM fl
  let h ← Meta.mkFreshExprMVar (Expr.const ``False [])
  let (_, mv) ← h.mvarId!.assertHypotheses is
  let (fvs, mv) ← mv.assertHypotheses ls
  mv.withContext do
    let hs := fvs.map (.fvar ·)
    _ ← Smt.smt mv hs.toList
    -- Note: auto should allow solvers to export new goals to users
    -- for mv in mvs do
    --   logInfo m!"new : {mv}"
    instantiateMVars h

attribute [rebind Auto.Native.solverFunc] smtSolverFunc
