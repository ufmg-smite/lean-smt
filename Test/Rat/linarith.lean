import Smt
import Smt.Rat

namespace Smt.Translate.Rat

open Qq
open Translator Term

@[smt_translate] def translateType : Translator := fun (e : Q(Type)) => match e with
  | ~q(Rat) => return symbolT "Real"
  | _       => return none

@[smt_translate] def translateRat : Translator := fun (e : Q(Rat)) => match e with
  | ~q(OfNat.ofNat $n) => return if let some n := n.rawNatLit? then literalT s!"{n}.0" else none
  | ~q(-$x)     => return appT (symbolT "-") (← applyTranslators! x)
  | ~q($x + $y) => return mkApp2 (symbolT "+") (← applyTranslators! x) (← applyTranslators! y)
  | ~q($x - $y) => return mkApp2 (symbolT "-") (← applyTranslators! x) (← applyTranslators! y)
  | ~q($x * $y) => return mkApp2 (symbolT "*") (← applyTranslators! x) (← applyTranslators! y)
  | ~q($x / $y) => return mkApp2 (symbolT "/") (← applyTranslators! x) (← applyTranslators! y)
  | _           => return none

@[smt_translate] def translateProp : Translator := fun (e : Q(Prop)) => match e with
  | ~q(($x : Rat) < $y) => return mkApp2 (symbolT "<") (← applyTranslators! x) (← applyTranslators! y)
  | ~q(($x : Rat) ≤ $y) => return mkApp2 (symbolT "<=") (← applyTranslators! x) (← applyTranslators! y)
  | ~q(($x : Rat) ≥ $y) => return mkApp2 (symbolT ">=") (← applyTranslators! x) (← applyTranslators! y)
  | ~q(($x : Rat) > $y) => return mkApp2 (symbolT ">") (← applyTranslators! x) (← applyTranslators! y)
  | _                   => return none

end Smt.Translate.Rat

example {x y : Rat} {f : Rat → Rat} : ¬(((1/2 : Rat) = 1/2) ∧ x ≤ y ∧ y ≤ x ∧ ¬f x = f y) := by
  smt

example {x y : Rat} {f : Rat → Rat} : ¬(((1/2 : Int) = 1/2) ∧ x ≤ y ∧ y ≤ x ∧ ¬f x = f y) := by
  smt

example (A B : Rat) (h : 0 < A * B) : 0 < A*B/8 := by
  smt [h]

example (A B : Rat) (h : 0 < A * B) : 0 < A/8*B := by
  smt [h]

example (ε : Rat) (h1 : ε > 0) : ε / 2 + ε / 3 + ε / 7 < ε := by
  smt [h1]

example (x y z : Rat) (h1 : 2*x < 3*y) (h2 : -4*x + z/2 < 0)
        (h3 : 12*y - z < 0)  : False := by
  smt [h1, h2, h3]

example (ε : Rat) (h1 : ε > 0) : ε / 2 < ε := by
  smt [h1]

example (ε : Rat) (h1 : ε > 0) : ε / 3 + ε / 3 + ε / 3 = ε := by
  smt [h1]

example (x : Rat) (h : 0 < x) : 0 < x/1 := by
  smt [h]

example (x : Rat) (h : x < 0) : 0 < x/(-1) := by
  smt [h]

example (x : Rat) (h : x < 0) : 0 < x/(-2) := by
  smt [h]

example (x : Rat) (h : x < 0) : 0 < x/(-4/5) := by
  smt [h]

example (x : Rat) (h : 0 < x) : 0 < x/2/3 := by
  smt [h]

example (x : Rat) (h : 0 < x) : 0 < x/(2/3) := by
  smt [h]

example (a b c : Rat) (h2 : b + 2 > 3 + b) : False := by
  smt [h2]

example (g v V c h : Rat) (h1 : h = 0) (h2 : v = V) (h3 : V > 0) (h4 : g > 0)
    (h5 : 0 ≤ c) (h6 : c < 1) : v ≤ V := by
  smt [h1, h2, h3, h4, h5, h6]

example (a b c : Rat) (h1 : a > 0) (h2 : b > 5) (h3 : c < -10) (h4 : a + b - c < 3) : False := by
  smt [h1, h2, h3, h4]

example (a b c : Rat) (h2 : b > 0) (h3 : ¬ b ≥ 0) : False := by
  smt [h2, h3]

example (x y z : Rat) (hx : x ≤ 3*y) (h2 : y ≤ 2*z) (h3 : x ≥ 6*z) : x = 3*y := by
  smt [hx, h2, h3]

example (x y z : Rat) (hx : ¬ x > 3*y) (h2 : ¬ y > 2*z) (h3 : x ≥ 6*z) : x = 3*y := by
  smt [hx, h2, h3]

-- example (x y : Rat) (h : 6 + ((x + 4) * x + (6 + 3 * y) * y) = 3) (h' : (x + 4) * x ≥ 0)
--     (h'' : (6 + 3 * y) * y ≥ 0) : False := by
--   smt [h, h'']
--   all_goals sorry

example (a : Rat) (ha : 0 ≤ a) : 0 * 0 ≤ 2 * a := by
  smt [ha]

example (x y : Rat) (h : x < y) : x ≠ y := by
  smt [h]

example (x y : Rat) (h : x < y) : ¬ x = y := by
  smt [h]

-- example (x : Rat) : id x ≥ x := by
--   let idℝ := fun (x : Rat) => x
--   let hidℝ : idℝ = fun x => x := rfl
--   have : @id Rat = idℝ := rfl
--   rewrite [this]
--   smt [idℝ]
--   all_goals (ring_nf; simp)
--   sorry

example (prime : Int → Prop) (x y z : Rat) (h1 : 2*x + ((-3)*y) < 0) (h2 : (-4)*x + 2*z < 0) (h3 : 12*y + (-4)* z < 0)
    (h4 : prime 7) : False := by
  smt [h1, h2, h3, h4]

example (prime : Int → Prop) (x y z : Rat) (h1 : 2*1*x + (3)*(y*(-1)) < 0) (h2 : (-2)*x*2 < -(z + z))
    (h3 : 12*y + (-4)* z < 0) (h4 : prime 7) : False := by
  smt [h1, h2, h3, h4]

example : ∃ (x : Rat), x > 2 ∧ x * x < 24 := by
  smt

example : ∃ (x : Rat), x > 2 ∧ x * 2 < 24 := by
  smt
