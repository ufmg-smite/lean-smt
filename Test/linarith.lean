import Smt

example {x y : Real} {f : Real → Real} : ¬(((1/2 : Real) = 1/2) ∧ x ≤ y ∧ y ≤ x ∧ ¬f x = f y) := by
  smt

example {x y : Real} {f : Real → Real} : ¬(((1/2 : Int) = 1/2) ∧ x ≤ y ∧ y ≤ x ∧ ¬f x = f y) := by
  smt

example {x y : Int} {f : Int → Int} : ¬(x ≤ y ∧ y ≤ x ∧ ¬f x = f y) := by
  smt

example {p q r : Prop} (hp : ¬p) (hq : ¬q) (hr : r) : ¬(p ∨ q ∨ ¬r) := by
  smt [hp, hq, hr]

example {p q r : Prop} : ((p ∧ q) ∧ r) = (r ∧ True ∧ q ∧ p ∧ p) := by
  smt

example {a b : Int} (h : a < b) (w : b < a) : False := by
  smt [h, w]

example
    {a b c : Int}
    (ha : a < 0)
    (hb : ¬b = 0)
    (hc' : c = 0)
    (h₁ : (1 - a) * (b * b) ≤ 0)
    (hc : (0 : Int) ≤ 0)
    (w : -(a * -b * -b + b * -b + 0) = (1 - a) * (b * b))
    (h₂ : (1 - a) * (b * b) ≤ 0) :
    0 < 1 - a := by
  smt [ha, hb, hc', h₁, hc, w, h₂]

example (e b c a v0 v1 : Int) (h1 : v0 = 5*a) (h2 : v1 = 3*b)
    (h3 : v0 + v1 + c = 10) : v0 + 5 + (v1 - 3) + (c - 2) = 10 := by
  smt [h1, h2, h3]

example (h : (1 : ℤ) < 0) (g : ¬ (37 : ℤ) < 42) (_k : True) (l : (-7 : ℤ) < 5): (3 : ℤ) < 7 := by
  smt [h, g, _k, l]

example (u v r s t : Int) (h : 0 < u*(t*v + t*r + s)) : 0 < (t*(r + v) + s)*3*u := by
  smt [h]

example (A B : Int) (h : 0 < 3 * A * B) : 0 < 8*A*B := by
  smt [h]

example (A B : Int) (h : 0 < 8 * A * B) : 0 < A*B := by
  smt [h]

example (A B : Int) (h : 0 < A * B) : 0 < A*8*B := by
  smt [h]

example (x : Int) : 0 ≤ x := by
  have h : 0 ≤ x := sorry
  smt [h]

example (u v r s t : Int) (h : 0 < u*(t*v + t*r + s)) :
    0 < (t*(r + v) + s)*3*u := by
  smt [h]

example (A B : Int) (h : 0 < A * B) : 0 < 8*A*B := by
  smt [h]

example (A B : Real) (h : 0 < A * B) : 0 < A*B/8 := by
  smt [h]
  all_goals sorry

example (A B : Real) (h : 0 < A * B) : 0 < A/8*B := by
  smt [h]
  all_goals sorry

example (ε : Real) (h1 : ε > 0) : ε / 2 + ε / 3 + ε / 7 < ε := by
  smt [h1]
  all_goals sorry

example (x y z : Real) (h1 : 2*x < 3*y) (h2 : -4*x + z/2 < 0)
        (h3 : 12*y - z < 0)  : False := by
  smt [h1, h2, h3]
  all_goals sorry

example (ε : Real) (h1 : ε > 0) : ε / 2 < ε := by
  smt [h1]
  all_goals sorry

example (ε : Real) (h1 : ε > 0) : ε / 3 + ε / 3 + ε / 3 = ε := by
  smt [h1]
  all_goals sorry

example (x : Real) (h : 0 < x) : 0 < x/1 := by
  smt [h]
  all_goals sorry

example (x : Real) (h : x < 0) : 0 < x/(-1) := by
  smt [h]
  all_goals (ring_nf; simp)

example (x : Real) (h : x < 0) : 0 < x/(-2) := by
  smt [h]
  all_goals (ring_nf; simp)

example (x : Real) (h : x < 0) : 0 < x/(-4/5) := by
  smt [h]
  all_goals (ring_nf; simp)

example (x : Real) (h : 0 < x) : 0 < x/2/3 := by
  smt [h]
  all_goals (ring_nf; simp)

example (x : Real) (h : 0 < x) : 0 < x/(2/3) := by
  smt [h]
  all_goals (ring_nf; simp)

example (a b c : Real) (h2 : b + 2 > 3 + b) : False := by
  smt [h2]
  all_goals (ring_nf; simp)
  norm_num

example (a b c : Real) (h2 : b + 2 > 3 + b) : False := by
  smt [h2]
  all_goals (ring_nf; simp)
  norm_num

example (g v V c h : Real) (h1 : h = 0) (h2 : v = V) (h3 : V > 0) (h4 : g > 0)
    (h5 : 0 ≤ c) (h6 : c < 1) : v ≤ V := by
  smt [h1, h2, h3, h4, h5, h6]

example (x y z : ℤ) (h1 : 2*x < 3*y) (h2 : -4*x + 2*z < 0) (h3 : 12*y - 4* z < 0) : False := by
  smt [h1, h2, h3]

example (x y z : ℤ) (h1 : 2*x < 3*y) (h2 : -4*x + 2*z < 0) (h3 : x*y < 5) (h3 : 12*y - 4* z < 0) :
    False := by
  smt [h1, h2, h3]

example (a b c : Real) (h1 : a > 0) (h2 : b > 5) (h3 : c < -10) (h4 : a + b - c < 3) : False := by
  smt [h1, h2, h3, h4]

example (a b c : Real) (h2 : b > 0) (h3 : ¬ b ≥ 0) : False := by
  smt [h2, h3]

example (x y z : Real) (hx : x ≤ 3*y) (h2 : y ≤ 2*z) (h3 : x ≥ 6*z) : x = 3*y := by
  smt [hx, h2, h3]

example (x y z : ℤ) (h1 : 2*x < 3*y) (h2 : -4*x + 2*z < 0) (h3 : x*y < 5) : ¬ 12*y - 4* z < 0 := by
  smt [h1, h2, h3]

example (x y z : Real) (hx : ¬ x > 3*y) (h2 : ¬ y > 2*z) (h3 : x ≥ 6*z) : x = 3*y := by
  smt [hx, h2, h3]

-- example (x y : Real) (h : 6 + ((x + 4) * x + (6 + 3 * y) * y) = 3) (h' : (x + 4) * x ≥ 0)
--     (h'' : (6 + 3 * y) * y ≥ 0) : False := by
--   smt [h, h'']
--   all_goals sorry

example (a : Real) (ha : 0 ≤ a) : 0 * 0 ≤ 2 * a := by
  smt [ha]

example (x y : Real) (h : x < y) : x ≠ y := by
  smt [h]

-- example (x y : Real) (h : x < y) : ¬ x = y := by
--   smt [h]

-- example (x : Real) : id x ≥ x := by
--   let idℝ := fun (x : Real) => x
--   let hidℝ : idℝ = fun x => x := rfl
--   have : @id Real = idℝ := rfl
--   rewrite [this]
--   smt [idℝ]
--   all_goals (ring_nf; simp)
--   sorry

example (prime : ℤ → Prop) (x y z : Real) (h1 : 2*x + ((-3)*y) < 0) (h2 : (-4)*x + 2*z < 0) (h3 : 12*y + (-4)* z < 0)
    (h4 : prime 7) : False := by
  smt [h1, h2, h3, h4]

example (prime : ℤ → Prop) (x y z : Real) (h1 : 2*1*x + (3)*(y*(-1)) < 0) (h2 : (-2)*x*2 < -(z + z))
    (h3 : 12*y + (-4)* z < 0) (h4 : prime 7) : False := by
  smt [h1, h2, h3, h4]

example (prime : ℤ → Prop) (w x y z : ℤ) (h1 : 4*x + (-3)*y + 6*w ≤ 0) (h2 : (-1)*x < 0) (h3 : y < 0) (h4 : w ≥ 0)
    (h5 : prime x) : False := by
  smt [h1, h2, h3, h4]

set_option maxRecDepth 200000 in
example (u v x y A B : Int)
(a : 0 < A)
(a_1 : 0 <= 1 - A)
(a_2 : 0 <= B - 1)
(a_3 : 0 <= B - x)
(a_4 : 0 <= B - y)
(a_5 : 0 <= u)
(a_6 : 0 <= v)
(a_7 : 0 < A - u)
(a_8 : 0 < A - v) :
 (0 < A * A)
-> (0 <= A * (1 - A))
-> (0 <= A * (B - 1))
-> (0 <= A * (B - x))
-> (0 <= A * (B - y))
-> (0 <= A * u)
-> (0 <= A * v)
-> (0 < A * (A - u))
-> (0 < A * (A - v))
-> (0 <= (1 - A) * A)
-> (0 <= (1 - A) * (1 - A))
-> (0 <= (1 - A) * (B - 1))
-> (0 <= (1 - A) * (B - x))
-> (0 <= (1 - A) * (B - y))
-> (0 <= (1 - A) * u)
-> (0 <= (1 - A) * v)
-> (0 <= (1 - A) * (A - u))
-> (0 <= (1 - A) * (A - v))
-> (0 <= (B - 1) * A)
-> (0 <= (B - 1) * (1 - A))
-> (0 <= (B - 1) * (B - 1))
-> (0 <= (B - 1) * (B - x))
-> (0 <= (B - 1) * (B - y))
-> (0 <= (B - 1) * u)
-> (0 <= (B - 1) * v)
-> (0 <= (B - 1) * (A - u))
-> (0 <= (B - 1) * (A - v))
-> (0 <= (B - x) * A)
-> (0 <= (B - x) * (1 - A))
-> (0 <= (B - x) * (B - 1))
-> (0 <= (B - x) * (B - x))
-> (0 <= (B - x) * (B - y))
-> (0 <= (B - x) * u)
-> (0 <= (B - x) * v)
-> (0 <= (B - x) * (A - u))
-> (0 <= (B - x) * (A - v))
-> (0 <= (B - y) * A)
-> (0 <= (B - y) * (1 - A))
-> (0 <= (B - y) * (B - 1))
-> (0 <= (B - y) * (B - x))
-> (0 <= (B - y) * (B - y))
-> (0 <= (B - y) * u)
-> (0 <= (B - y) * v)
-> (0 <= (B - y) * (A - u))
-> (0 <= (B - y) * (A - v))
-> (0 <= u * A)
-> (0 <= u * (1 - A))
-> (0 <= u * (B - 1))
-> (0 <= u * (B - x))
-> (0 <= u * (B - y))
-> (0 <= u * u)
-> (0 <= u * v)
-> (0 <= u * (A - u))
-> (0 <= u * (A - v))
-> (0 <= v * A)
-> (0 <= v * (1 - A))
-> (0 <= v * (B - 1))
-> (0 <= v * (B - x))
-> (0 <= v * (B - y))
-> (0 <= v * u)
-> (0 <= v * v)
-> (0 <= v * (A - u))
-> (0 <= v * (A - v))
-> (0 < (A - u) * A)
-> (0 <= (A - u) * (1 - A))
-> (0 <= (A - u) * (B - 1))
-> (0 <= (A - u) * (B - x))
-> (0 <= (A - u) * (B - y))
-> (0 <= (A - u) * u)
-> (0 <= (A - u) * v)
-> (0 < (A - u) * (A - u))
-> (0 < (A - u) * (A - v))
-> (0 < (A - v) * A)
-> (0 <= (A - v) * (1 - A))
-> (0 <= (A - v) * (B - 1))
-> (0 <= (A - v) * (B - x))
-> (0 <= (A - v) * (B - y))
-> (0 <= (A - v) * u)
-> (0 <= (A - v) * v)
-> (0 < (A - v) * (A - u))
-> (0 < (A - v) * (A - v))
->
 u * y + v * x + u * v < 3 * A * B := by
  smt [a, a_1, a_2, a_3, a_4, a_5, a_6, a_7, a_8]
