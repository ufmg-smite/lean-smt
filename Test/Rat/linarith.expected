Test/Rat/linarith.lean:74:9: warning: unused variable `a`
note: this linter can be disabled with `set_option linter.unusedVariables false`
Test/Rat/linarith.lean:74:13: warning: unused variable `c`
note: this linter can be disabled with `set_option linter.unusedVariables false`
Test/Rat/linarith.lean:84:9: warning: unused variable `a`
note: this linter can be disabled with `set_option linter.unusedVariables false`
Test/Rat/linarith.lean:84:13: warning: unused variable `c`
note: this linter can be disabled with `set_option linter.unusedVariables false`
